From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Milan Zamazal <mzamazal@redhat.com>
Date: Wed, 21 Oct 2026 09:59:45 +0200
Subject: [PATCH 07/10] ipa: softisp: Add Af algorithm

Let's add Af algorithm, which is mostly a proxy to the libipa Af.

The only special feature is setting afEnabled flag to prevent attempting
lens movements when the algorithm is not enabled.

The algorithm is not enabled yet because there are no autofocus stats
computed.

Signed-off-by: Milan Zamazal <mzamazal@redhat.com>
---
 src/ipa/simple/algorithms/af.cpp      | 53 +++++++++++++++++++++++++++
 src/ipa/simple/algorithms/af.h        | 39 ++++++++++++++++++++
 src/ipa/simple/algorithms/meson.build |  1 +
 src/ipa/simple/data/uncalibrated.yaml |  1 +
 src/ipa/simple/ipa_context.h          |  4 ++
 5 files changed, 98 insertions(+)
 create mode 100644 src/ipa/simple/algorithms/af.cpp
 create mode 100644 src/ipa/simple/algorithms/af.h

diff --git a/src/ipa/simple/algorithms/af.cpp b/src/ipa/simple/algorithms/af.cpp
new file mode 100644
index 00000000..a2a21632
--- /dev/null
+++ b/src/ipa/simple/algorithms/af.cpp
@@ -0,0 +1,53 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+/*
+ * Copyright (C) 2026 Red Hat Inc.
+ *
+ * Autofocus
+ */
+
+#include "af.h"
+
+#include <libcamera/base/log.h>
+
+#include <libcamera/control_ids.h>
+
+#include "libcamera/internal/software_isp/swisp_stats.h"
+
+namespace libcamera {
+
+namespace ipa::soft::algorithms {
+
+LOG_DEFINE_CATEGORY(IPASoftAf)
+
+int Af::init(IPAContext &context, [[maybe_unused]] const ValueNode &tuningData)
+{
+	/* Informs that lens controls can get focus data */
+	context.afEnabled = true;
+
+	return 0;
+}
+
+int Af::configure(IPAContext &context,
+		  [[maybe_unused]] const IPAConfigInfo &configInfo)
+{
+	afAlgo_.configure(context.activeState.af);
+
+	return 0;
+}
+
+void Af::process(IPAContext &context,
+		 [[maybe_unused]] const uint32_t frame,
+		 [[maybe_unused]] IPAFrameContext &frameContext,
+		 const SwIspStats *stats,
+		 [[maybe_unused]] ControlList &metadata)
+{
+	const double variance = stats->variance;
+	LOG(IPASoftAf, Debug) << "Focus stats: " << variance;
+	afAlgo_.process(context.activeState.af, variance);
+}
+
+REGISTER_IPA_ALGORITHM(Af, "Af")
+
+} /* namespace ipa::soft::algorithms */
+
+} /* namespace libcamera */
diff --git a/src/ipa/simple/algorithms/af.h b/src/ipa/simple/algorithms/af.h
new file mode 100644
index 00000000..9b04ad2c
--- /dev/null
+++ b/src/ipa/simple/algorithms/af.h
@@ -0,0 +1,39 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+/*
+ * Copyright (C) 2026 Red Hat Inc.
+ *
+ * Autofocus
+ */
+
+#pragma once
+
+#include "libipa/af.h"
+
+#include "algorithm.h"
+#include "ipa_context.h"
+
+namespace libcamera {
+
+namespace ipa::soft::algorithms {
+
+class Af : public Algorithm
+{
+public:
+	Af() = default;
+	~Af() = default;
+
+	int init(IPAContext &context, const ValueNode &tuningData) override;
+	int configure(IPAContext &context,
+		      const IPAConfigInfo &configInfo) override;
+	void process(IPAContext &context, const uint32_t frame,
+		     IPAFrameContext &frameContext,
+		     const SwIspStats *stats,
+		     ControlList &metadata) override;
+
+private:
+	AfAlgorithm afAlgo_;
+};
+
+} /* namespace ipa::soft::algorithms */
+
+} /* namespace libcamera */
diff --git a/src/ipa/simple/algorithms/meson.build b/src/ipa/simple/algorithms/meson.build
index 73c63722..1cd58d5e 100644
--- a/src/ipa/simple/algorithms/meson.build
+++ b/src/ipa/simple/algorithms/meson.build
@@ -2,6 +2,7 @@
 
 soft_simple_ipa_algorithms = files([
     'adjust.cpp',
+    'af.cpp',
     'awb.cpp',
     'agc.cpp',
     'blc.cpp',
diff --git a/src/ipa/simple/data/uncalibrated.yaml b/src/ipa/simple/data/uncalibrated.yaml
index fc90ca52..53d3792a 100644
--- a/src/ipa/simple/data/uncalibrated.yaml
+++ b/src/ipa/simple/data/uncalibrated.yaml
@@ -4,6 +4,7 @@
 version: 1
 algorithms:
   - BlackLevel:
+  # - Af:
   - Awb:
   # Color correction matrices can be defined here. The CCM algorithm
   # has a significant performance impact, and should only be enabled
diff --git a/src/ipa/simple/ipa_context.h b/src/ipa/simple/ipa_context.h
index 8ccfacb4..556f036e 100644
--- a/src/ipa/simple/ipa_context.h
+++ b/src/ipa/simple/ipa_context.h
@@ -17,6 +17,7 @@
 #include "libcamera/internal/vector.h"
 
 #include <libipa/fc_queue.h>
+#include "libipa/af.h"
 
 #include "core_ipa_interface.h"
 
@@ -36,6 +37,8 @@ struct IPASessionConfiguration {
 };
 
 struct IPAActiveState {
+	ipa::af::ActiveState af;
+
 	struct {
 		int32_t exposure;
 		double again;
@@ -90,6 +93,7 @@ struct IPAContext {
 	FCQueue<IPAFrameContext> frameContexts;
 	ControlInfoMap::Map ctrlMap;
 	bool ccmEnabled = false;
+	bool afEnabled = false;
 };
 
 } /* namespace ipa::soft */
